packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,186 @@
+// @ts-check
+
+// no-lonely-if is a stupid rule that really should be disabled globally
+/* eslint-disable no-lonely-if */
+
+import { assert } from '@agoric/assert';
+import { M } from '@agoric/store';
+import { parseVatSlot } from '../lib/parseVatSlots.js';
+
+export function makeWatchedPromiseManager(
+  syscall,
+  vrm,
+  vom,
+  cm,
+  convertValToSlot,
+  convertSlotToVal,
+) {
+  const { makeScalarBigMapStore } = cm;
+  const { defineDurableKind } = vom;
+
+  // watched promises by vpid: each entry is an array of watches on the
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        Promise.resolve().then(() => {
+          if (wasFulfilled) {
+            if (watcher.onFulfilled) {
+              watcher.onFulfilled(value, ...args);
+            }
+          } else {
+            if (watcher.onRejected) {
+              watcher.onRejected(value, ...args);
+            } else {
+              throw value;
+            }
+          }
+        });
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises');
+    let deadPromises;
+    if (deadPromisesRaw) {
+      deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+    } else {
+      deadPromises = new Set();
+    }
+    for (const [vpid, watches] of watchedPromiseTable.entries()) {
+      if (deadPromises.has(vpid)) {
+        watchedPromiseTable.delete(vpid);
+        for (const watch of watches) {
+          const [watcher, ...args] = watch;
+          if (watcher.onRejected) {
+            Promise.resolve().then(() => {
+              watcher.onRejected('vat upgraded', ...args);
+            });
+          }
+        }
+      } else {
+        const p = convertSlotToVal(vpid);
```

Comment:
This `convertSlotToVal` is an interesting case: when version-2 starts, the kernel will see a duplicate `syscall.subscribe()` for each not-dead yes-watched promise `vpid`, because this is normally the pathway by which the vat imports promises that arrive in message arguments. If the kernel-side promise is unresolved, the `syscall.subscribe` is a no-op (`kernelKeeper.addSubscriberToPromise` uses a `Set` for deduplication). If the promise already was resolved (perhaps it was resolved just before the vat is upgraded, but the `notify` is still sitting in a queue), the kernel will enqueue a duplicate `notify` event, however that will be discarded when it arrives at the vat and we see the `vpid` missing from the c-list.

So, not a problem, but something to keep in mind.

Reply:
For the record: per our conversations we've replaced this use of `convertSlotToVal` with a new liveslots function, `revivePromise`, which is expected to be used in this context and avoids the various landmines.